### PR TITLE
End Chat Optional Parameter

### PR DESCRIPTION
**Pull request by elopezanaya**
# PR Details

### Description
_Include a description of the problem to be solved_

For sessions that were disconnected , or ended by agent, when calling endchat is forcing an exception since the conversation is already closed.

## Solution Proposed
_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_

Adding optional parameters to allow the caller to identify when a call to end chat is for an already closed chat, this will avoid to call close chat endpoint.

 #### this PR includes a cleanup of warnings and style issues in unit tests :  
- extra lines
- trail spaces
- unused variables
- rename let to const
- whitelist for empty arrow functions
- whitelist for any 
- whitelist unused declared variables inside callbacks
- fix tabs and spaces

### Acceptance criteria
_Define what are the conditions to consider the PR has achieved the intended goal_

## Test cases and evidence

image

_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_
image

### Sanity Tests

- [ ] You have tested all changes in Popout mode
- [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
- [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y

- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

***Please provide justification if any of the validations has been skipped.***

### CODE REVIEW

File: src/OmnichannelChatSDK.ts [hunk 1]
```diff
@@ -779,38 +780,60 @@ class OmnichannelChatSDK {
...
+            ChatId: this.chatToken.chatId as string,
+        };
 
-        const sessionCloseOptionalParams: ISessionCloseOptionalParams = {};
+        // in case a session was ended by agent or disconnected, there is no need to close the session
+        this.scenarioMarker.startScenario(TelemetryEvent.CloseChatSession, cleanupMetadata);
 
-        if (this.isPersistentChat && !this.chatSDKConfig.persistentChat?.disable) {
-            const isReconnectChat = this.reconnectId !== null ? true : false;
+        if (!endChatOptionalParams?.isSessionEnded) {
```
Comment (mithunjohny) [inline src/OmnichannelChatSDK.ts:793]: should we add isSessionEnded to telemetry so that it is easier to distinguish between customer and agent

File: src/core/EndChatOptionalParams.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+export default interface EndChatOptionalParams {
+isSessionEnded?: false;
```
Comment (xTEddie) [inline src/core/EndChatOptionalParams.ts:3]: nit: spacing

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 3]
```diff
@@ -3131,6 +3130,99 @@ describe('Omnichannel Chat SDK, Sequential', () => {
             expect(chatSDK.IC3Client).toBe(undefined);
         });
 
+        it('ChatSDK.endChat() should end conversation without calling closeSession', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:3133]: `ChatSDK.endChat() with optional params \'isSessionEnded\' set as \'true\' should not call sessionClose`

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 4]
```diff
@@ -3131,6 +3130,99 @@ describe('Omnichannel Chat SDK, Sequential', () => {
...
+            });
+
+            expect(chatSDK.OCClient.sessionClose).toHaveBeenCalledTimes(0);
+            expect(conversationDisconnectFn).toHaveBeenCalledTimes(1);
+            expect(chatSDK.conversation).toBe(null);
+            expect(chatSDK.chatToken).toMatchObject({});
+            expect(chatSDK.IC3Client).toBe(undefined);
+        });
+
+        it('ChatSDK.endChat() should end conversation passing optional params', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:3164]: `ChatSDK.endChat() with optional params \'isSessionEnded\' set as \'false\' should call sessionClose`

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 5]
```diff
@@ -3131,6 +3130,99 @@ describe('Omnichannel Chat SDK, Sequential', () => {
...
+            });
+
+            expect(chatSDK.OCClient.sessionClose).toHaveBeenCalledTimes(1);
+            expect(conversationDisconnectFn).toHaveBeenCalledTimes(1);
+            expect(chatSDK.conversation).toBe(null);
+            expect(chatSDK.chatToken).toMatchObject({});
+            expect(chatSDK.IC3Client).toBe(undefined);
+        });
+
+        it('ChatSDK.endChat() should end conversation passing optional params as undefined', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:3195]: `ChatSDK.endChat() with optional params \'isSessionEnded\' set as \'undefined\' should call sessionClose`

File: playwright/integrations/unauthenticated-chat.spec.ts [hunk 6]
```diff
@@ -541,6 +542,40 @@ test.describe('UnauthenticatedChat @UnauthenticatedChat', () => {
         expect(sessionCloseResponse.status()).toBe(200);
     });
 
+    test('ChatSDK.endChat() should not call close session for identified sessions ended by agent', async ({ page }) => {
```
Comment (xTEddie) [inline playwright/integrations/unauthenticated-chat.spec.ts:545]: `ChatSDK.endChat() with optional params \'isSessionEnded\' set as \'true\' should call not sessionClose`

File: __tests__/OmnichannelChatSDKParallel.spec.ts [hunk 7]
```diff
@@ -2374,6 +2374,158 @@ describe('Omnichannel Chat SDK, Parallel initialization', () => {
             expect(chatSDK.IC3Client).toBe(undefined);
         });
 
+        it('ChatSDK.endChat() should end conversation without calling backend close session', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDKParallel.spec.ts:2377]: 'ChatSDK.endChat() with optional params \'isSessionEnded\' set as \'true\' should not call sessionClose'

File: __tests__/OmnichannelChatSDKParallel.spec.ts [hunk 8]
```diff
@@ -2374,6 +2374,158 @@ describe('Omnichannel Chat SDK, Parallel initialization', () => {
...
+            });
+
+            expect(chatSDK.OCClient.sessionClose).toHaveBeenCalledTimes(0);
+            expect(conversationDisconnectFn).toHaveBeenCalledTimes(1);
+            expect(chatSDK.conversation).toBe(null);
+            expect(chatSDK.chatToken).toMatchObject({});
+            expect(chatSDK.IC3Client).toBe(undefined);
+        });
+
+        it('ChatSDK.endChat() should end conversation not ended by agent', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDKParallel.spec.ts:2415]: 'ChatSDK.endChat() with optional params \'isSessionEnded\' set as \'false\' should call sessionClose'

File: __tests__/OmnichannelChatSDKParallel.spec.ts [hunk 9]
```diff
@@ -2374,6 +2374,158 @@ describe('Omnichannel Chat SDK, Parallel initialization', () => {
...
+
+            expect(chatSDK.OCClient.sessionClose).toHaveBeenCalledTimes(1);
+            expect(conversationDisconnectFn).toHaveBeenCalledTimes(1);
+            expect(chatSDK.conversation).toBe(null);
+            expect(chatSDK.chatToken).toMatchObject({});
+            expect(chatSDK.IC3Client).toBe(undefined);
+        });
+
+        
+        it('ChatSDK.endChat() should end conversation not ended by agent and empty optional params', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDKParallel.spec.ts:2454]: 'ChatSDK.endChat() with optional params set as empty should call sessionClose'

File: __tests__/OmnichannelChatSDKParallel.spec.ts [hunk 10]
```diff
@@ -2374,6 +2374,158 @@ describe('Omnichannel Chat SDK, Parallel initialization', () => {
...
+
+            expect(chatSDK.OCClient.sessionClose).toHaveBeenCalledTimes(1);
+            expect(conversationDisconnectFn).toHaveBeenCalledTimes(1);
+            expect(chatSDK.conversation).toBe(null);
+            expect(chatSDK.chatToken).toMatchObject({});
+            expect(chatSDK.IC3Client).toBe(undefined);
+        });
+
+
+        it('ChatSDK.endChat() should end conversation not ended by agent and not defined in optional params', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDKParallel.spec.ts:2491]: 'ChatSDK.endChat() with optional params set as 'undefined' should call sessionClose'

File: __tests__/OmnichannelChatSDKParallel.spec.ts [hunk 11]
```diff
@@ -2465,6 +2617,42 @@ describe('Omnichannel Chat SDK, Parallel initialization', () => {
             expect(chatSDK.OCClient.sessionClose.mock.calls[0][1]).toMatchObject(sessionCloseOptionalParams);
         });
 
+        it('ChatSDK.endChat() should end conversation not ended by agent and passing empty object', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDKParallel.spec.ts:2620]: 'ChatSDK.endChat() with optional params set as empty should call sessionClose'

**Comment by elopezanaya** [comment]
fixing multiple style issues in unit tests, no affectation or changes in logic

**Review by matan-hershberg-nuance — CHANGES_REQUESTED** [review]

File: __tests__/OmnichannelChatSDKParallel.spec.ts [hunk 12]
```diff
@@ -2374,6 +2374,158 @@ describe('Omnichannel Chat SDK, Parallel initialization', () => {
...
+        it('ChatSDK.endChat() should end conversation without calling backend close session', async () => {
+            const chatSDK = new OmnichannelChatSDK(omnichannelConfig);
+            chatSDK.getChatConfig = jest.fn();
+            chatSDK.getChatToken = jest.fn();
+
+            await chatSDK.initialize({ useParallelLoad: true });
+            let retryCount = 0;
+            const maxRetries = 3;
+
+            while (chatSDK.AMSClient === null && retryCount < maxRetries) {
```
[the remark this answers is not in the document]
Comment (elopezanaya) [reply __tests__/OmnichannelChatSDKParallel.spec.ts:2370]: this is out of the scope, is just testing the call to end chat , so Im reusing existent test with the modification for the input parameter,

File: src/OmnichannelChatSDK.ts [hunk 13]
```diff
@@ -779,38 +780,64 @@ class OmnichannelChatSDK {
...
-        }
+                const sessionCloseOptionalParams: ISessionCloseOptionalParams = {};
 
-        if (this.isChatReconnect && !this.chatSDKConfig.chatReconnect?.disable && !this.isPersistentChat) {
-            const isChatReconnect = this.reconnectId !== null ? true : false;
-            this.requestId = isChatReconnect ? (this.reconnectId as string) : this.requestId; // Chat Reconnect session to close
-            sessionCloseOptionalParams.isReconnectChat = isChatReconnect;
-        }
+                if (this.isPersistentChat && !this.chatSDKConfig.persistentChat?.disable) {
+                    const isReconnectChat = this.reconnectId !== null ? true : false;
```
[the remark this answers is not in the document]
Comment (elopezanaya) [reply src/OmnichannelChatSDK.ts:800]: same, Im just reusing existent tests, doing cleanup, not need to enhance test logic

**Review by matan-hershberg-nuance — APPROVED** [review]

**Review by mithunjohny — APPROVED** [review]
